tables/AlternateMans/BufferedColumnarFile.h

```diff
@@ -0,0 +1,307 @@
+ *
+ * This class limits calls to the read/write functions by caching a block of
+ * data around the last accessed row. The size of the buffer is templated to
+ * make testing of the caching easier. Production code can normally use the
+ * alias BufferedColumnarFile, which uses a reasonable buffer size.
+ *
+ * This class aims to be as fast as possible for large data files that are read
+ * from or written to consecutively. It does not try to optimize random access,
+ * partial data access and access is not transactional.
+ *
+ * The class uses exceptions to handle any I/O errors.
+ */
+template <uint64_t BufferSize = 100 * 1024>
+class VarBufferedColumnarFile : private RowBasedFile {
+ public:
+  using RowBasedFile::AddRows;
+  using RowBasedFile::DeleteRow;
+  using RowBasedFile::Filename;
+  using RowBasedFile::IsOpen;
+  using RowBasedFile::NRows;
+  using RowBasedFile::ReadHeader;
+  using RowBasedFile::Stride;
+  using RowBasedFile::WriteHeader;
+
+  VarBufferedColumnarFile() noexcept = default;
+
+  VarBufferedColumnarFile(const VarBufferedColumnarFile& rhs) = delete;
+  VarBufferedColumnarFile(VarBufferedColumnarFile&& rhs) noexcept
+      : packed_buffer_(std::move(rhs.packed_buffer_)),
+        block_changed_(rhs.block_changed_),
+        active_block_(rhs.active_block_),
+        rows_per_block_(rhs.rows_per_block_),
+        block_buffer_(std::move(rhs.block_buffer_)) {
+    rhs.block_changed_ = false;
+    rhs.active_block_ = 0;
+    rhs.rows_per_block_ = 0;
+  }
+
+  ~VarBufferedColumnarFile() noexcept {
+    if (IsOpen()) {
+      if (block_changed_) {
+        const uint64_t start_row = active_block_ * rows_per_block_;
+        const size_t n_rows_to_write =
+            std::min(rows_per_block_, std::max(NRows(), start_row) - start_row);
+        Seek(start_row * Stride() + DataLocation(), SEEK_SET);
+        WriteData(block_buffer_.data(), n_rows_to_write * Stride());
+        block_changed_ = false;
+      }
+    }
+  }
+
+  VarBufferedColumnarFile& operator=(VarBufferedColumnarFile&& rhs) {
+    RowBasedFile::operator=(std::move(rhs));
+    std::swap(packed_buffer_, rhs.packed_buffer_);
+    std::swap(block_changed_, rhs.block_changed_);
+    std::swap(active_block_, rhs.active_block_);
+    std::swap(rows_per_block_, rhs.rows_per_block_);
+    std::swap(block_buffer_, rhs.block_buffer_);
+    return *this;
+  }
+
+  /**
+   * Close the file. After closing, all calls to I/O functions
+   * cause undefined behaviour, untill the class is assigned to
+   * a new instance.
+   */
+  void Close() {
+    if (IsOpen()) {
+      if (block_changed_) {
+        const uint64_t start_row = active_block_ * rows_per_block_;
+        const size_t n_rows_to_write =
+            std::min(rows_per_block_, std::max(NRows(), start_row) - start_row);
+        Seek(start_row * Stride() + DataLocation(), SEEK_SET);
+        WriteData(block_buffer_.data(), n_rows_to_write * Stride());
+        block_changed_ = false;
+      }
```

Comment:
Can you create a (private) function for this code, since it's now duplicated in the destructor, here and in `ActivateBlock`.

Reply:
Good idea, done